### PR TITLE
Sparse matrices and sparse_dot product (as a module)

### PR DESCRIPTION
**Pull request by yger**
The sparse matrix is now an external module, that can be loaded separatly such that cudamat does not depend on scipy. There is also a test for sparse_dot product.

### CODE REVIEW

**Comment by yger** [comment]
Sorry for the fact that the beginning of the branch is messy, only the last commits are relevant. I moved everything for sparse matrices into a distinct sparse.py file, and added test/docstrings. Please let me know if I can do more. As I may continue to work on cudamat, this would be easier to merge our work. I think this is a great package!

**Comment by f0k** [comment]
Hey, thanks a lot for reviving this, and sorry for the late reply! It currently conflicts with the master branch so it cannot be merged. Can you try to fix the conflicts? The way to go would be:
```bash
git checkout -b backup sparse  # create a backup of your current branch
git fetch upstream master  # download the latest upstream version
git checkout sparse
git rebase upstream/master  # rebase your branch onto the upstream version
```
There will be a conflict. You will need to edit the conflicted file, use `git add` on the file and `git rebase --continue` to get through. If everything looks good, do `git push --force` to update this PR, no need to start a new one.

Also it would be nice to clean up the history. You can run:
```bash
git rebase -i upstream/master
```
for an interactive rebase. It allows you to merge existing commits into one. If you don't know what to do, merging all commits into a single file would probably be okay for this PR. Again, `git push --force` updates the PR.

Please let me know when you've cleaned up things (or when you have trouble), github doesn't automatically give a notification for that!

**Comment by yger** [comment]
Ok, I did as you suggested, and now this should be safe and clean for merge

File: cudamat/cudamat.cu [hunk 1]
```diff
@@ -425,6 +493,7 @@ EXPORT int add_col_vec(cudamat* mat, cudamat* vec, cudamat* target) {
     return 0;
 }
 
+
```
Comment (f0k) [inline cudamat/cudamat.cu]: please remove this unrelated whitespace change

File: cudamat/cudamat.py [hunk 2]
```diff
@@ -6,7 +6,6 @@
 
 import ctypes as ct
 import numpy as np
-
```
Comment (f0k) [inline cudamat/cudamat.py]: please remove this unrelated whitespace change

File: cudamat/cudamat.py [hunk 3]
```diff
@@ -94,6 +93,7 @@ def load_library(basename):
 _cudamat.vdot.restype = ct.c_float
 _cudamat.dot.restype = ct.c_int
 
+
```
Comment (f0k) [inline cudamat/cudamat.py]: please remove this unrelated whitespace change

File: cudamat/cudamat.py [hunk 4]
```diff
@@ -196,6 +196,10 @@ def __init__(self, mat):
         self.p_mat = ct.pointer(self.mat)
 
 
+
+
+
+
```
Comment (f0k) [inline cudamat/cudamat.py]: please remove this unrelated whitespace change

File: cudamat/cudamat.py [hunk 5]
```diff
@@ -235,6 +239,9 @@ def __init__(self, array, copy_to_device=True, copy_on_host=True):
                 if err_code:
                     raise generate_exception(err_code)
 
+            if not copy_on_host:
+                del self.numpy_array
+
```
Comment (f0k) [inline cudamat/cudamat.py]: This is not what `copy_on_host` means -- it's fully covered by the `copy=copy_on_host` flag in `reformat` above. It is just a way to avoid an unneeded extra copy on the host, not a way to avoid linking the matrix to a numpy array at all. The extra copy is still enabled by default to not break backwards compatibility.

File: cudamat/cudamat.py [hunk 6]
```diff
@@ -1274,6 +1281,7 @@ def mean(mat, axis, target=None):
     return sum(mat, axis, target=target, mult=1. / mat.shape[axis])
 
 
+
```
Comment (f0k) [inline cudamat/cudamat.py]: please remove this unrelated whitespace change

File: cudamat/cudamat_kernels.cu [hunk 7]
```diff
@@ -782,11 +799,9 @@ __global__ void kWhere(float* condition_mat, float* if_mat, float* else_mat, flo
     }
 }
 
-
```
Comment (f0k) [inline cudamat/cudamat_kernels.cu]: please remove this unrelated whitespace change

File: cudamat/cudamat_kernels.cu [hunk 8]
```diff
@@ -782,11 +799,9 @@ __global__ void kWhere(float* condition_mat, float* if_mat, float* else_mat, flo
...
 __global__ void kCorrelate(float* source, float* kernel, float* dest, int width, int height, int kwidth, int kheight) {
     const unsigned int idx = blockIdx.x * blockDim.x + threadIdx.x;
     const unsigned int numThreads = blockDim.x * gridDim.x;
-
```
Comment (f0k) [inline cudamat/cudamat_kernels.cu:789]: please remove this unrelated whitespace change

File: cudamat/cudamat_kernels.cuh [hunk 9]
```diff
@@ -89,4 +90,4 @@ __global__ void kSelectRows(float* source, float* target, float* indices, int nR
 __global__ void kSetSelectedRows(float* target, float* source, float* indices, int nRowIs, int nCols, int nTargetRows);
 __global__ void kWhere(float* condition_mat, float* if_mat, float* else_mat, float* target, unsigned int len);
 __global__ void kCorrelate(float* source, float* kernel, float* dest, int width, int height, int fwidth, int fheight);
-#endif
+#endif
```
Comment (f0k) [inline cudamat/cudamat_kernels.cuh:94]: please remove this unrelated whitespace change

File: cudamat/sparse.py [hunk 10]
```diff
@@ -0,0 +1,101 @@
...
+    precision floats on a GPU.
+    """
+    def __init__(self, array, copy_to_device = True, copy_on_host=True):
+        """
+        Initializes a new matrix object in one of two ways. If array is a numpy
+        ndarray, memory for a matrix with the same dimensions is allocated on
+        the GPU. If the copy_to_device flag is set to True, the GPU matrix is
+        initialized with the given ndarray. If array is not an ndarray, it must
+        be a cudamat structure (typically the user will never use this way of
+        calling __init__).
```
Comment (f0k) [inline cudamat/sparse.py]: Needs updating. `array` is not a `ndarray` but a scipy CSR matrix, and `cuamat` structure is a `cudamat_sparse` structure. Furthermore, `copy_on_host` is not explained.

File: cudamat/sparse.py [hunk 11]
```diff
@@ -0,0 +1,101 @@
...
+        """
+
+        assert(type(array) == sp.csr_matrix)
+        self.mat = cudamat_sparse()
+        self.size = self.mat.size
+        self.p_mat = ct.pointer(self.mat)
+        self.scipy_array = array.astype('float32')
```
Comment (f0k) [inline cudamat/sparse.py]: `astype` always creates a copy. We don't want a copy if `copy_on_host` is falseish.

**Comment by yger** [comment]
I did changes, as suggested. Hope this is now better

**Comment by yger** [comment]
What can I do more, such that this is merged in the main branch? Please let me know, I may try to add some more functions after, always with tests and being more cautious about whitespace changes :-)

**Comment by f0k** [comment]
Sorry for the long delay, I've been busy... and thank you for addressing the changes I've noted so far. I will finish reviewing the code later this week!

**Comment by f0k** [comment]
Sorry again, will have to defer this until after the ICLR deadline (Nov 4). I haven't forgotten!
